### PR TITLE
feat(request-client.js): add erc20 proxy contract PN in request-client.js

### PR DESCRIPTION
**Pull request by ltacker**
## Description of the changes

Add function in `request-client.js` to create a request with the erc20 proxy contract payment network.
`getBalance` function is not yet implemented

## Link to Jira

https://requestnetwork.atlassian.net/browse/PROT-1015

### CODE REVIEW

File: packages/request-client.js/test/api/payment-network/erc20/proxy-contract.test.ts [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+        return;
+      },
+      createCreationAction(): any {
+        return;
+      },
+    },
+  },
+};
+
+// Most of the tests are done as integration tests in ../index.test.ts
```
Comment (benjlevesque) [inline packages/request-client.js/test/api/payment-network/erc20/proxy-contract.test.ts]: ?

File: packages/request-client.js/src/api/payment-network/erc20/proxy-contract.ts [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+   * Will set a salt if none is already given
+   *
+   * @param paymentNetworkCreationParameters Parameters to create the extension
+   * @returns The extensionData object
+   */
+  public createExtensionsDataForCreation(
+    paymentNetworkCreationParameters: ExtensionTypes.PnReferenceBased.ICreationParameters,
+  ): ExtensionTypes.IAction {
+    // If no salt is given, generate one
+    const salt = paymentNetworkCreationParameters.salt || Utils.crypto.generate8randomBytes();
```
Comment (benjlevesque) [inline packages/request-client.js/src/api/payment-network/erc20/proxy-contract.ts]: should this be tested more specifically? like a test that validates that `createdCreationAction` is called with a salt, even if no salt was passed as parameters

File: packages/request-client.js/test/api/payment-network/erc20/proxy-contract.test.ts [hunk 3]
```diff
@@ -0,0 +1,90 @@
+import { AdvancedLogicTypes, RequestLogicTypes } from '@requestnetwork/types';
+import ERC20ProxyContract from '../../../../src/api/payment-network/erc20/proxy-contract';
+
+import 'chai';
+import 'mocha';
+
+const chai = require('chai');
```
Comment (benjlevesque) [inline packages/request-client.js/test/api/payment-network/erc20/proxy-contract.test.ts]: use `import` ? :)

File: packages/request-client.js/src/api/payment-network/erc20/proxy-contract.ts [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+/**
+ * Handle payment networks with ERC20 proxy contract extension
+ */
+export default class PaymentNetworkERC20ProxyContract implements Types.IPaymentNetwork {
+  private extension: ExtensionTypes.PnReferenceBased.IReferenceBased;
+  /**
+   * @param extension The advanced logic payment network extensions
+   */
+  public constructor({ advancedLogic }: { advancedLogic: AdvancedLogicTypes.IAdvancedLogic }) {
+    this.extension = advancedLogic.extensions.erc20ProxyContract;
```
Comment (benjlevesque) [inline packages/request-client.js/src/api/payment-network/erc20/proxy-contract.ts]: not familiar with how this works, but maybe we should test that it's defined here?

File: packages/request-client.js/test/api/payment-network/erc20/proxy-contract.test.ts (hunk 1 above)
Comment (ltacker) [reply packages/request-client.js/test/api/payment-network/erc20/proxy-contract.test.ts]: Most of the features of the payment network are tested in the file `index.test.ts` by creating directly a request with the payment network.
Good point from this comment: those tests are missing

Comment (ltacker) [reply packages/request-client.js/test/api/payment-network/erc20/proxy-contract.test.ts]: In fact for this payment network it's not most test

File: packages/request-client.js/src/api/payment-network/erc20/proxy-contract.ts (hunk 4 above)
Comment (ltacker) [reply packages/request-client.js/src/api/payment-network/erc20/proxy-contract.ts]: extension is a private member, that's why we doesn't test it but your comment made me creating this task: https://requestnetwork.atlassian.net/browse/PROT-1047

File: packages/request-client.js/src/api/payment-network/erc20/proxy-contract.ts [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+   * Will set a salt if none is already given
+   *
+   * @param paymentNetworkCreationParameters Parameters to create the extension
+   * @returns The extensionData object
+   */
+  public createExtensionsDataForCreation(
+    paymentNetworkCreationParameters: Types.IReferenceBasedCreationParameters,
+  ): ExtensionTypes.IAction {
+    // If no salt is given, generate one
+    const salt = paymentNetworkCreationParameters.salt || Utils.crypto.generate8randomBytes();
```
Comment (vrolland) [inline packages/request-client.js/src/api/payment-network/erc20/proxy-contract.ts]: `Utils.crypto.generate8randomBytes();`
Is it the right format? it should be string in 'hex', no? @romaric-juniet ?
if the function returns such a string the naming is bad (and it's probably my bad)

File: packages/request-client.js/test/api/payment-network/erc20/proxy-contract.test.ts [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+  it('can createExtensionsDataForCreation', async () => {
+    const spy = sandbox.on(mockAdvancedLogic.extensions.proxyContractErc20, 'createCreationAction');
+
+    erc20ProxyContract.createExtensionsDataForCreation({
+      paymentAddress: 'ethereum address',
+      salt: 'ea3bc7caf64110ca',
+    });
+
+    expect(spy).to.have.been.called.once;
```
Comment (vrolland) [inline packages/request-client.js/test/api/payment-network/erc20/proxy-contract.test.ts]: this doesn't test much. We should at least check the argument given to the funcion called.

File: packages/request-client.js/test/api/payment-network/erc20/proxy-contract.test.ts [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+  });
+
+  it('can createExtensionsDataForCreation without salt', async () => {
+    const spy = sandbox.on(mockAdvancedLogic.extensions.proxyContractErc20, 'createCreationAction');
+
+    erc20ProxyContract.createExtensionsDataForCreation({
+      paymentAddress: 'ethereum address',
+    });
+
+    expect(spy).to.have.been.called.once;
```
Comment (vrolland) [inline packages/request-client.js/test/api/payment-network/erc20/proxy-contract.test.ts]: Same here.

File: packages/request-client.js/test/api/payment-network/erc20/proxy-contract.test.ts [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+    const spy = sandbox.on(
+      mockAdvancedLogic.extensions.proxyContractErc20,
+      'createAddPaymentAddressAction',
+    );
+
+    erc20ProxyContract.createExtensionsDataForAddPaymentInformation({
+      paymentAddress: 'ethereum address',
+    });
+
+    expect(spy).to.have.been.called.once;
```
Comment (vrolland) [inline packages/request-client.js/test/api/payment-network/erc20/proxy-contract.test.ts]: same

File: packages/request-client.js/test/api/payment-network/erc20/proxy-contract.test.ts [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+    const spy = sandbox.on(
+      mockAdvancedLogic.extensions.proxyContractErc20,
+      'createAddRefundAddressAction',
+    );
+
+    erc20ProxyContract.createExtensionsDataForAddRefundInformation({
+      refundAddress: 'ethereum address',
+    });
+
+    expect(spy).to.have.been.called.once;
```
Comment (vrolland) [inline packages/request-client.js/test/api/payment-network/erc20/proxy-contract.test.ts]: and... same :)

File: packages/request-client.js/test/api/payment-network/erc20/proxy-contract.test.ts (hunk 6 above)
Comment (ltacker) [reply packages/request-client.js/test/api/payment-network/erc20/proxy-contract.test.ts]: The problem is for all PN tests, so I created this task:
https://requestnetwork.atlassian.net/secure/RapidBoard.jspa?rapidView=22&projectKey=PROT&view=planning&selectedIssue=PROT-1056

Also chai seems to not support `to.be.called.with()` with object with parameter, it seems we need to use sinon package instead.

File: packages/request-client.js/src/api/payment-network/erc20/proxy-contract.ts (hunk 5 above)
Comment (romaric-juniet) [reply packages/request-client.js/src/api/payment-network/erc20/proxy-contract.ts]: image

Comment (romaric-juniet) [reply packages/request-client.js/src/api/payment-network/erc20/proxy-contract.ts]: it generates 8 random bytes and returns them as a string

File: packages/request-client.js/test/api/payment-network/erc20/proxy-contract.test.ts (hunk 6 above)
Comment (vrolland) [reply packages/request-client.js/test/api/payment-network/erc20/proxy-contract.test.ts]: Discussed IRL. 
chai support it but the error message is not helping to understand what happened.

**Pull request merged**